rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,226 @@
+
+- Cache key: The unified cache key can be composed of two parts: backend specific
+  key and backend id. The backend id is unique among backends. The backend
+  specific key can be defined in each backends, to achieve more flexibility.
+- Cache value: The unified cache value can be a base buffer class like following,
+  and each backend can inherit this base class and subclass it.
+
+```cpp
+class constant_buffer_t{
+public:
+    virtual size_t get_size() const;
+    virtual ~constant_buffer_t();
+
+private:
+    void *data_;
+    size_t size_;
+    // Memory freeing relies on engine and allocator
+    const dnnl::engine eng_;
+    const allocator_t *alc_;
+};
+```
+
+The constant tensor cache utils may looks like the following. It will be
+templated class which can be specialized and instantiated for each engine kind
+for dedication purpose. The support for concurrent execution is same as the
+existing DNNL backend constant tensor cache, which used the promise-future
+mechanism. That is not changed.
+
+```cpp
+template<class engine_kind>
+class unified_constant_tensor_cache_t {
+public:
+    using key_t = size_t;
+    using cached_t = std::shared_ptr<constant_buffer_t>;
+    using value_t = std::shared_future<cached_t>;
+    using promise_t = std::promise<cached_t>;
+
+    unified_constant_tensor_cache_t();
+    ~unified_constant_tensor_cache_t();
+
+    status_t set_capacity(size_t capacity);
+    size_t get_capacity();
+
+    /// @backend_id:
+    /// @backend_specific_key:
+    /// @size: The size the constant tensor
+    /// @value: A promised future constant tensor. If cache miss, the thread
+    /// which firstly query the constant tensor will promise to compute it in
+    /// the future, then other threads who also query the same constant tensor
+    /// will wait util the first thread ready.
+    /// @return: A valid future of cache hit, and an invalid one if cache miss.
+    value_t get_or_add(size_t backend_id, size_t backend_specific_key,
+            sze_t size, const value_t &value);
+    void remove_if_exist(size_t backend_id, size_t backend_specific_key);
+};
+
+template<typename engine_kind>
+unified_constant_tensor_cache_t &get_constant_tensor_cache();
+```
+
+The `set_capacity` and `get_capacity` method will be wrapped to frontend API,
+and the `get_or_add` and `remove_if_exist` method will be provided to backend as
+a part of backend API. The `get_or_add` method could be used when a backend want
+to query a constant tensor from the cache. If cache miss, the backend will add a
+promised future value to the cache and fulfill the promise after the tensor
+computation completes (The mentioned promise-future mechanism). The
+`remove_if_exist` method could be used when backend want to delete an item from
+the cache.
+
+When the cached tensor size reach the capacity, the new coming tensor won't be
+cached anymore by default. In later section, we will further discuss the policy
```

Comment:
How to ensure that `When the cached tensor size reach the capacity, the new coming tensor won't be cached anymore by default`? What will happen if backend still executes `get_or_add` when cached tensor size reaches the capacity?

Reply:
It's ensured by the constant tensor cache itself. backend won't be aware of this and could still executes `get_or_add` as normal. When cache miss, the future value passed in through this API won't be added to cache

Reply:
`When cache miss, the future value passed in through this API won't be added to cache`, will this API return any status for this case? Otherwise, backend will always run in a path like query->miss->add->fail->query->miss->add->fail which is not inefficient.

Reply:
the `get_or_add` api only return status to show cache miss or cache hit, it doesn't tell why cache miss (first time to access the tensor or reach capacity). backend could query the capacity and current cache size by using other APIs to deduce the reason and do more optimizations.